aws/resource_aws_security_group_test.go

```diff
@@ -35,93 +35,54 @@ func init() {
 
 func testSweepSecurityGroups(region string) error {
 	client, err := sharedClientForRegion(region)
+
 	if err != nil {
 		return fmt.Errorf("error getting client: %s", err)
 	}
+
 	conn := client.(*AWSClient).ec2conn
+	sweepResources := make([]*testSweepResource, 0)
+	var errs *multierror.Error
 
 	input := &ec2.DescribeSecurityGroupsInput{}
 
 	// Delete all non-default EC2 Security Group Rules to prevent DependencyViolation errors
 	err = conn.DescribeSecurityGroupsPages(input, func(page *ec2.DescribeSecurityGroupsOutput, lastPage bool) bool {
+		if page == nil {
+			return !lastPage
+		}
+
 		for _, sg := range page.SecurityGroups {
 			if aws.StringValue(sg.GroupName) == "default" {
 				log.Printf("[DEBUG] Skipping default EC2 Security Group: %s", aws.StringValue(sg.GroupId))
 				continue
 			}
 
-			if sg.IpPermissions != nil {
-				req := &ec2.RevokeSecurityGroupIngressInput{
-					GroupId:       sg.GroupId,
-					IpPermissions: sg.IpPermissions,
-				}
-
-				if _, err = conn.RevokeSecurityGroupIngress(req); err != nil {
-					log.Printf("[ERROR] Error revoking ingress rule for Security Group (%s): %s", aws.StringValue(sg.GroupId), err)
-				}
-			}
-
-			if sg.IpPermissionsEgress != nil {
-				req := &ec2.RevokeSecurityGroupEgressInput{
-					GroupId:       sg.GroupId,
-					IpPermissions: sg.IpPermissionsEgress,
-				}
+			r := resourceAwsSecurityGroup()
+			d := r.Data(nil)
+			d.SetId(aws.StringValue(sg.GroupId))
+			d.Set("revoke_rules_on_delete", true)
 
-				if _, err = conn.RevokeSecurityGroupEgress(req); err != nil {
-					log.Printf("[ERROR] Error revoking egress rule for Security Group (%s): %s", aws.StringValue(sg.GroupId), err)
-				}
-			}
+			sweepResources = append(sweepResources, NewTestSweepResource(r, d, client))
 		}
 
 		return !lastPage
 	})
 
-	if testSweepSkipSweepError(err) {
-		log.Printf("[WARN] Skipping EC2 Security Group sweep for %s: %s", region, err)
-		return nil
-	}
-
 	if err != nil {
-		return fmt.Errorf("Error retrieving EC2 Security Groups: %s", err)
+		errs = multierror.Append(errs, fmt.Errorf("error describing EC2 Security Groups for %s: %w", region, err))
 	}
 
-	err = conn.DescribeSecurityGroupsPages(input, func(page *ec2.DescribeSecurityGroupsOutput, lastPage bool) bool {
```

Comment:
If I recall correctly, the previous behavior was necessary as you could have security groups that cross-reference each other so it was not possible to delete them with revoking rules first.

Reply:
The resource itself does all that as far as I can tell when `d.Set("revoke_rules_on_delete", true)`.

Reply:
The problem isn't with the resource being deleted, it is a race condition that _other_ Security Groups may have rules that reference the current Security Group, which would cause the failures.